src/dataflow-types/src/client.rs

```diff
@@ -658,6 +667,120 @@ pub mod partitioned {
 
     use timely::progress::frontier::MutableAntichain;
 
+    /// Buffer for partial results for a `TAIL`.
+    /// This exists so we can consolidate rows and
+    /// sort them by timestamp before passing them to the consumer
+    /// (currently, coord.rs).
+    struct PendingTail {
+        /// `frontiers[i]` is an antichain representing the times at which we may
+        /// still receieve results from worker `i`.
+        frontiers: HashMap<usize, Antichain<Timestamp>>,
+        /// The results (possibly unsorted) which have not yet been delivered.
+        buffer: Vec<(Row, Timestamp, Diff)>,
+        /// The number of unique shard IDs expected.
+        parts: usize,
+        /// The last progress message that has been reported to the consumer.
+        reported_frontier: Antichain<Timestamp>,
+    }
+
+    impl PendingTail {
+        // See [`differential_dataflow::consolidation::consolidate_updates_slice`].
+        // This function works similarly, except that we sort by timestamp first, then data.
+        //
+        // I implemented it without `unsafe`; we can revisit that if proves to be
+        // a hot path.
+        /// Return all the updates with time less than `upper` (or all times if that is `None`),
+        /// in sorted and consolidated representation.
+        pub fn consolidate_up_to(
+            &mut self,
+            upper: Antichain<Timestamp>,
```

Comment:
Maybe silly, but if we stored everything as `(Time, Row, Diff)` could we just use `consolidate`?

Reply:
`consolidate` should work on any `(Ord, Ord, Semigroup)` types, so maybe it works directly here? But it's probably a good idea to stick to `(Time, Data, Diff)` to avoid confusion.

Reply:
@frankmcsherry I considered that, but it would require an extra operator to shuffle everything -- I wasn't sure if that was worth avoiding having to write this code.

@antiguru the reason I can't use `consolidate_updates` directly is because we want to sort by `time` followed by `data` here, rather than `data` followed by `time` as that function does.